internal/driver/glfw/loop.go

```diff
@@ -27,35 +26,50 @@ type drawData struct {
 	done chan struct{} // Zero allocation signalling channel
 }
 
+type runFlag struct {
+	sync.RWMutex
+	flag bool
+	cond *sync.Cond
+}
+
 // channel for queuing functions on the main thread
 var funcQueue = make(chan funcData)
 var drawFuncQueue = make(chan drawData)
-var runFlag uint32 = 0 // atomic, 0 == stopped, 1 == running
+var run *runFlag
 var initOnce = &sync.Once{}
 var donePool = &sync.Pool{New: func() interface{} {
 	return make(chan struct{})
 }}
 
+func newRun() *runFlag {
+	r := runFlag{}
+	r.cond = sync.NewCond(&r)
+	return &r
+}
+
 // Arrange that main.main runs on main thread.
 func init() {
 	runtime.LockOSThread()
-}
 
-func running() bool {
-	return atomic.LoadUint32(&runFlag) == 1
+	run = newRun()
 }
 
 // force a function f to run on the main thread
 func runOnMain(f func()) {
 	// If we are on main just execute - otherwise add it to the main queue and wait.
 	// The "running" variable is normally false when we are on the main thread.
-	if !running() {
+	run.RLock()
+	if !run.flag {
 		f()
```

Comment:
I used a RLock to minimize the lock cost (no contention once main is running), but it will be the equivalent of an compare and swap twice. Definitively more costly than one atomic read (Not by much, it still likely only trash cache across CPU only once like atomic), but I don't think we can get away from it.